opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/RuntimeEnvironment.java

```diff
@@ -1996,14 +1843,45 @@ public File getDtagsEftar() {
     }
 
     public SuggesterConfig getSuggesterConfig() {
-        return (SuggesterConfig) getConfigurationValue("suggesterConfig");
+        return syncReadConfiguration(Configuration::getSuggesterConfig);
     }
 
     public void setSuggesterConfig(SuggesterConfig config) {
-        setConfigurationValue("suggesterConfig", config);
+        syncWriteConfiguration(Configuration::setSuggesterConfig, config);
+    }
+
+    /**
+     * Applies the specified function to the runtime configuration, after having
+     * obtained the configuration read-lock (and releasing afterward).
+     * @param function a defined function
+     * @param <R> the type of the result of the function
+     * @return the function result
+     */
+    public <R> R syncReadConfiguration(Function<Configuration, R> function) {
+        try (ResourceLock resourceLock = configLock.readLockAsResource()) {
+            //noinspection ConstantConditions to silence unreference auto-closeable
+            assert resourceLock != null;
+            return function.apply(configuration);
+        }
+    }
+
+    /**
+     * Performs the specified operation which is provided the runtime
+     * configuration and the specified argument, after first having obtained the
+     * configuration write-lock (and releasing afterward).
+     * @param consumer a defined consumer
+     * @param v the input argument
+     * @param <V> the type of the input to the operation
+     */
+    public <V> void syncWriteConfiguration(ConfigurationValueConsumer<V> consumer, V v) {
```

Comment:
I'd swap the parameter order cause at it seems more natural to me like this

```
syncWriteConfiguration(groups, (c, g) -> {
      populateGroups(g, new TreeSet<>(getProjects().values()));
      c.setGroups(g);
});
```

as its "value -> consume" that value, rather then "consumer of a value -> value"

Reply:
That would mean a line like the following:

`syncWriteConfiguration(flag, Configuration::setListDirsFirst);`

which seems to invert the important part.

Reply:
I'll rename the vague arguments and transpose it as you suggest

Reply:
thank you